ltm/agent.py

```diff
@@ -96,6 +125,121 @@ def sanitize_string(s: str) -> str:
+
+            if triggered:
+                extra_tasks.append(task["task_details"])
+
+        return extra_tasks
+
+    def add_task(self, context, cost_callback, user_message): # given some trigger condition
+        task_should_add = """Here is the current list of future tasks:
+{future_tasks}
+
+----------
+You are to decide if this message:
+{user_message}
+ 
+Contains instructions for a future task which is not already in the list of future tasks, and has very concrete conditions (either a number of messages, a specific date/time, or a 'trigger phrase').
+- Does the description of the task imply that the thing that the agent has to do something given some future trigger?
+- Does the user tell the agent that it should perform some kind of task in the future, along with some trigger condition for that task?
+- Is the task already in the list?
+
+Respond with your reasoning as to whether there has been a new future task discussed and with either "Should Update: Yes." or "Should Update: No." written exactly at the very end of your response."""
+
+        new_task_def = """Create a new task in JSON format.
+A task contains two things, a trigger condition and a payload.
+
+Be very precise in how you define the task and record all relevant information about it including:
+- The trigger condition should describe precisely and only the condition for the triggering of the task.
+- The payload should describe precisely and only what should be done once the trigger condition is met.
+- The trigger value should accurately state the value stated in the trigger condition.
+
+Follow this format:
+{
+    "description": str, // A high level description of the task
+    "task_details": {
+        "trigger_condition": str // A precise description of the condition that should be met for the task to execute.
+        "payload": str, // An exact description of what should be done when the trigger condition is met. Do not mention the trigger
+        // Any other details that might be relevant
+    }
+    "trigger_details": {
+        "trigger_condition": str, // Either MESSAGE_COUNT, MESSAGE_CONTENT, or TIME
+        "trigger_value": str, // For MESSAGE_COUNT, the number of messages to wait. For TIME, a timestamp in the form `YYYY-mm-dd_HH:MM`. For MESSAGE_CONTENT, the exact text that will trigger the task.
+    }
+}"""
+
+        context.append(make_user_message(task_should_add.format(future_tasks=json.dumps(self.task_memory, indent=2), user_message=user_message)))
+        response_text = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+        log_llm_call(self.run_name, self.save_name, self.debug_level, label=f"should_update-{self.llm_call_idx}")
+        colour_print("YELLOW", f"Should update: {response_text}")
+
+        if "Should Update: No." in response_text:
+            return
+        elif "Should Update: Yes." not in response_text:
+            return
```

Comment:
This seems a bit too rigid. I usually ask the LLM to give me a single-word 'yes' or 'no' answer, and then I just do `if "no" in response_text.lower()` to ignore possible casing or formatting variations.

Reply:
Here the rigidity is the point. The agent has a bunch of free text to think though the points and a  look for "no" matches a ton of words. The alternative is to have it answer in JSON.

Reply:
Ah right. So the point is to have the LLM write some reasoning and let that decision be the last part of it 👌🏾